client/pages/sign-in.jsx

```diff
@@ -0,0 +1,154 @@
+import React from 'react';
+import GlobalContext from '../lib/global-context';
+
+export default class SignIn extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      username: '',
+      password: '',
+      passwordType: 'password',
+      usernameTooLong: false,
+      invalidLogin: false
+    };
+    this.handleChange = this.handleChange.bind(this);
+    this.handleSubmit = this.handleSubmit.bind(this);
+    this.togglePassword = this.togglePassword.bind(this);
+  }
+
+  handleChange(event) {
+    const { name, value } = event.target;
+    let newValue = value;
+    let usernameTooLong = false;
+    if (name === 'username' && newValue.length > 16) {
+      newValue = this.state.username;
+      usernameTooLong = true;
+    }
+    this.setState({ [name]: newValue, usernameTooLong });
+  }
+
+  handleSubmit(event) {
+    event.preventDefault();
+    const { username, password } = this.state;
+
+    let invalidLogin = false;
+    if (username.length < 4) {
+      invalidLogin = true;
+    }
+    if (password.length < 6) {
+      invalidLogin = true;
+    }
+    if (invalidLogin) {
+      this.setState({ invalidLogin });
+      return;
+    }
```

Comment:
This stuff shouldn't really be here. It's not common to put explicit error messages about the "shape" of the credentials.